File: go/internal/config/catalog.go
package config

import (
	"bytes"
	"crypto/sha256"
	"encoding/json/v2"
	"fmt"
	"io"
	"net/url"
	"os"
	"strings"

	"github.com/zR-JB/graphite-meter/go/internal/wire"
)

const maxCatalogBytes = 64 << 10

func loadServerCatalog() (wire.ServerCatalog, error) {
	raw, inline := os.LookupEnv("GM_SERVER_CATALOG")
	path, file := os.LookupEnv("GM_SERVER_CATALOG_FILE")
	if inline && file {
		return wire.ServerCatalog{}, fmt.Errorf("set only one of GM_SERVER_CATALOG and GM_SERVER_CATALOG_FILE")
	}
	if !inline && !file {
		return wire.SingletonCatalog(), nil
	}
	data := []byte(raw)
	if file {
		f, err := os.Open(path)
		if err != nil {
			return wire.ServerCatalog{}, fmt.Errorf("GM_SERVER_CATALOG_FILE: %w", err)
		}
		defer f.Close()
		data, err = io.ReadAll(io.LimitReader(f, maxCatalogBytes+1))
		if err != nil {
			return wire.ServerCatalog{}, err
		}
	}
	if len(data) > maxCatalogBytes {
		return wire.ServerCatalog{}, fmt.Errorf("server catalogue exceeds 64 KiB")
	}
	var c wire.ServerCatalog
	if bytes.HasPrefix(bytes.TrimSpace(data), []byte("[")) {
		var origins []string
		if err := json.Unmarshal(data, &origins); err != nil {
			return c, fmt.Errorf("server catalogue origins: %w", err)
		}
		for _, raw := range origins {
			canonical, err := wire.CanonicalOrigin(strings.TrimSuffix(raw, "/"))
			if err != nil {
				return c, fmt.Errorf("server catalogue origin %q: %w", raw, err)
			}
			u, _ := url.Parse(canonical)
			// The complete canonical origin owns identity, so reordering cannot redirect a saved selection.
			digest := sha256.Sum256([]byte(canonical))
			id := fmt.Sprintf("server-%x", digest[:16])
			c.Servers = append(c.Servers, wire.ServerEntry{ID: id, URL: canonical, Name: u.Host})
		}
	} else if err := json.Unmarshal(data, &c, json.RejectUnknownMembers(true)); err != nil {
		return c, fmt.Errorf("server catalogue: %w", err)
	}
	for i := range c.Servers {
		if c.Servers[i].ID == "self" {
			return c, fmt.Errorf("self is added automatically; omit it from servers")
		}
		canonical, err := wire.CanonicalOrigin(strings.TrimSuffix(c.Servers[i].URL, "/"))
		if err != nil {
			return c, fmt.Errorf("server %q: %w", c.Servers[i].ID, err)
		}
		c.Servers[i].URL = canonical
		for j, raw := range c.Servers[i].AdditionalOrigins {
			canonical, err := wire.CanonicalOrigin(strings.TrimSuffix(raw, "/"))
			if err != nil {
				return c, err
			}
			c.Servers[i].AdditionalOrigins[j] = canonical
		}
	}
	c.Servers = append(wire.SingletonCatalog().Servers, c.Servers...)
	if c.DefaultSelection == nil {
		c.DefaultSelection = []string{"self"}
	}
	if data, err := json.Marshal(c); err != nil || len(data) > 48<<10 {
		return c, fmt.Errorf("normalized catalogue exceeds 48 KiB; reserve space for this server's transport origins")
	}
	return c, c.Validate()
}

File: rust/server/src/catalog.rs
//! Bounded server catalogue configuration decoding.

use graphite_meter_core::{
    catalog::{ServerCatalog, ServerEntry},
    origin::{canonical_origin, target_origin},
};
use serde::Deserialize;
use sha2::{Digest, Sha256};
use std::{error::Error, fs::File, io::Read};

const MAX_INPUT_BYTES: usize = 64 << 10;
const MAX_NORMALIZED_BYTES: usize = 48 << 10;
type Result<T> = std::result::Result<T, Box<dyn Error + Send + Sync>>;

#[derive(Default, Deserialize)]
#[serde(default, deny_unknown_fields, rename_all = "camelCase")]
struct RawCatalog {
    default_selection: Option<Vec<Option<String>>>,
    servers: Option<Vec<Option<RawEntry>>>,
}

#[derive(Default, Deserialize)]
#[serde(default, deny_unknown_fields, rename_all = "camelCase")]
struct RawEntry {
    id: Option<String>,
    url: Option<String>,
    name: Option<String>,
    location: Option<String>,
    additional_origins: Option<Vec<Option<String>>>,
}

/// Source presence is significant, including empty strings. No process environment is read.
pub fn load(inline: Option<&str>, file: Option<&str>) -> Result<ServerCatalog> {
    match (inline, file) {
        (Some(_), Some(_)) => {
            Err("set only one of GM_SERVER_CATALOG and GM_SERVER_CATALOG_FILE".into())
        }
        (None, None) => Ok(ServerCatalog::singleton()),
        (Some(raw), None) => parse(raw.as_bytes()),
        (None, Some(path)) => {
            let mut data = Vec::new();
            File::open(path)?
                .take((MAX_INPUT_BYTES + 1) as u64)
                .read_to_end(&mut data)?;
            parse(&data)
        }
    }
}

fn canonical(raw: &str) -> Result<String> {
    Ok(canonical_origin(raw.strip_suffix('/').unwrap_or(raw))?)
}

pub fn parse(data: &[u8]) -> Result<ServerCatalog> {
    if data.len() > MAX_INPUT_BYTES {
        return Err("server catalogue exceeds 64 KiB".into());
    }
    let mut catalog = ServerCatalog::singleton();
    if data.iter().find(|b| !b.is_ascii_whitespace()) == Some(&b'[') {
        let origins: Vec<Option<String>> = serde_json::from_slice(data)?;
        for raw in origins {
            let url = canonical(&raw.unwrap_or_default())?;
            let origin = target_origin(&url)?.ok_or("expected absolute origin")?;
            let digest = Sha256::digest(url.as_bytes());
            let mut id = String::from("server-");
            use std::fmt::Write;
            for byte in &digest[..16] {
                write!(&mut id, "{byte:02x}")?;
            }
            catalog.servers.push(ServerEntry {
                id,
                url,
                name: origin.authority(),
                ..ServerEntry::default()
            });
        }
    } else {
        // Go accepts a root null as an empty catalogue, then inserts self.
        let raw: Option<RawCatalog> = serde_json::from_slice(data)?;
        let raw = raw.unwrap_or_default();
        if let Some(selected) = raw.default_selection {
            catalog.default_selection = selected
                .into_iter()
                .map(Option::unwrap_or_default)
                .collect();
        }
        for entry in raw.servers.unwrap_or_default() {
            let entry = entry.unwrap_or_default();
            let id = entry.id.unwrap_or_default();
            if id == "self" {
                return Err("self is added automatically; omit it from servers".into());
            }
            let url = canonical(&entry.url.unwrap_or_default())?;
            let additional_origins = entry
                .additional_origins
                .unwrap_or_default()
                .into_iter()
                .map(|raw| canonical(&raw.unwrap_or_default()))
                .collect::<Result<Vec<_>>>()?;
            catalog.servers.push(ServerEntry {
                id,
                url,
                name: entry.name.unwrap_or_default(),
                location: entry.location.unwrap_or_default(),
                additional_origins,
            });
        }
    }
    if serde_json::to_vec(&catalog)?.len() > MAX_NORMALIZED_BYTES {
        return Err("normalized catalogue exceeds 48 KiB; reserve space for this server's transport origins".into());
    }
    catalog.validate()?;
    Ok(catalog)
}
